assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java

```diff
@@ -3766,56 +3773,67 @@ private List getPrevFeedbackAttachments(Map<String, String> p) {
     /**
      * build the instructor preview of grading submission
      */
-    private String build_instructor_preview_grade_submission_context(VelocityPortlet portlet, Context context, RunData data,
-                                                                     SessionState state) {
+    private String build_instructor_preview_grade_submission_context(VelocityPortlet portlet, Context context, RunData data, SessionState state) {
 
-        // assignment
-        Assignment.GradeType gradeType = GRADE_TYPE_NONE;
-        String assignmentId = (String) state.getAttribute(GRADE_SUBMISSION_ASSIGNMENT_ID);
-        Assignment a = getAssignment(assignmentId, "build_instructor_preview_grade_submission_context", state);
-        if (a != null) {
-            context.put("assignment", a);
-            gradeType = a.getTypeOfGrade();
-        }
-
-        // submission
         AssignmentSubmission submission = getSubmission((String) state.getAttribute(GRADE_SUBMISSION_SUBMISSION_ID), "build_instructor_preview_grade_submission_context", state);
-        context.put("submission", submission);
+        if (submission != null) {
+            context.put("submission", submission);
 
-        if (a != null) {
-            setScoringAgentProperties(context, a, submission, false);
-        }
+            Map<String, Reference> submissionAttachmentReferences = new HashMap<>();
+            submission.getAttachments().forEach(r -> submissionAttachmentReferences.put(r, entityManager.newReference(r)));
+            context.put("submissionAttachmentReferences", submissionAttachmentReferences);
 
-        User user = (User) state.getAttribute(STATE_USER);
-        context.put("user", user);
-        context.put("submissionTypeTable", submissionTypeTable());
-        context.put("contentTypeImageService", contentTypeImageService);
+            Assignment assignment = submission.getAssignment();
+            context.put("assignment", assignment);
+
+            Map<String, Reference> assignmentAttachmentReferences = new HashMap<>();
+            assignment.getAttachments().forEach(r -> assignmentAttachmentReferences.put(r, entityManager.newReference(r)));
+            context.put("assignmentAttachmentReferences", assignmentAttachmentReferences);
 
-        // filter the feedback text for the instructor comment and mark it as red
-        String feedbackText = (String) state.getAttribute(GRADE_SUBMISSION_FEEDBACK_TEXT);
-        context.put("feedback_comment", state.getAttribute(GRADE_SUBMISSION_FEEDBACK_COMMENT));
-        context.put("feedback_text", feedbackText);
-        context.put("feedback_attachment", state.getAttribute(GRADE_SUBMISSION_FEEDBACK_ATTACHMENT));
+            StringBuilder submitterNames = new StringBuilder();
+            submission.getSubmitters().forEach(s -> {
+                try {
+                    User user = userDirectoryService.getUser(s.getSubmitter());
```

Comment:
Maybe just a TODO, but it's a lot faster if you can getUsers() so that LDAP can batch retrieve

Reply:
good tip thx